client/clientservice/src/state_snapshot_service.cpp

```diff
@@ -22,27 +27,405 @@ using vmware::concord::client::statesnapshot::v1::StreamSnapshotResponse;
+          timeout = curr_timeout;
+        }
+      }
+      if (!isHashValid(proto_request->snapshot_id(), accumulated_hash, timeout, status)) {
+        status = grpc::Status(grpc::StatusCode::FAILED_PRECONDITION, "Hash mismatch from servers");
+      }
+    }
+
+  } else {
+    status = grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Invalid argument sent");
+  }
+
+  return status;
+}
+
+bool StateSnapshotServiceImpl::isHashValid(uint64_t snapshot_id,
+                                           concord::util::SHA3_256::Digest& final_hash,
+                                           const std::chrono::milliseconds& timeout,
+                                           Status& return_status) {
+  ReadConfig read_config{RequestConfig{false, 0, 1024 * 1024, timeout, "signedHashReq", "", false, true},
+                         bft::client::ByzantineSafeQuorum{}};
+  std::vector<std::unique_ptr<SignedPublicStateHashResponse>> signed_hash_responses;
+  auto callback =
+      [&signed_hash_responses, &return_status, &read_config](concord::client::concordclient::SendResult&& send_result) {
+        ResponseType<SignedPublicStateHashResponse> reponse;
+        getResponseSetStatus(std::move(send_result),
+                             reponse,
+                             return_status,
+                             read_config.request.correlation_id,
+                             true,
+                             "concord.client.clientservice.state_snapshot_service.signedhash.callback");
+        signed_hash_responses =
+            std::get<std::vector<std::unique_ptr<SignedPublicStateHashResponse>>>(std::move(reponse.response));
+      };
+
+  ReconfigurationRequest rreq;
+  SignedPublicStateHashRequest cmd;
+  cmd.snapshot_id = snapshot_id;
+  cmd.participant_id = client_->getTRId();
+  rreq.command = cmd;
+  bft::client::Msg message;
+  concord::messages::serialize(message, rreq);
+  client_->send(read_config, std::move(message), callback);
+  for (const auto& res : signed_hash_responses) {
+    if (return_status.ok()) {
+      switch (res->status) {
+        case SnapshotResponseStatus::InternalError:
+          return_status = grpc::Status(grpc::StatusCode::UNKNOWN, "Internal error");
+          break;
+        case SnapshotResponseStatus::SnapshotNonExistent:
+          return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot doesn't exist");
+          break;
+        case SnapshotResponseStatus::SnapshotPending:
+          return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot not ready");
+          break;
+        case SnapshotResponseStatus::Success:
+          if ((res->data).snapshot_id != snapshot_id) {
+            return_status = grpc::Status(grpc::StatusCode::UNAVAILABLE, "Snapshot id mismatch");
+          }
+          concord::util::SHA3_256::Digest replica_hash;
+          for (size_t i = 0; i < (res->data).hash.size(); ++i) {
+            replica_hash[i] = ((res->data).hash)[i];
+          }
```

Comment:
I was wondering about the following here:

1. Aren't read-only reconfiguration responses containing data only in the replica-specific portion part (even without the signatures)? Or if the responses are the same (due to lack of signature), reconfiguration would put them in the common part?
2. Nit: Can we avoid copying the hash just to compare it?

Reply:
Thank you!